### PR TITLE
[patch] Rewrite Types Section

### PR DESCRIPTION
**Pull request by mmaloney-sf**
This PR is a moderate rewrite of the entire Types section.

This is best reviewed by building the PDF and reviewing paragraph by paragraph.

Here are some highlights:

- Uninferred types are mentioned in their own section. (Both width and reset inference).
- The section on zero-width integers is abbreviated. (It warrants mention, but doesn't need so much ink spilled)
- Analog type is scooched down.
- Small code formatting of FIRRTL (eg, spaces between braces in bundle types)
- Reword section on flips with bundle types.
- Rename "Reference Types" to "Probe Types"
- Move several examples of FIRRTL programs using probes out of the Probe Types section and into the probes section.

### CODE REVIEW

File: spec.md
Comment (seldridge) [inline spec.md:573]: The PR should be careful to avoid putting ABI information into the spec. E.g., this discussion could go in the direction of what happens to zero-width ports. However, that is part of the ABI guarantees of the port lowering ABIs.

There is something needed here, though!

File: spec.md
Comment (seldridge) [inline spec.md:582]: Is `conditionals` the correct section link?

File: spec.md
Comment (seldridge) [inline spec.md:611]:
```suggestion
Note that the element type of a vector can be any type, including another aggregate type.
```

File: spec.md
Comment (seldridge) [inline spec.md:627]: Mega-nit: I grok this is an intuitional sentence. However, `set` makes me ask if this is a mathematical set and whether it is ordered (it is from the port-lowering ABI). Is there a better way of stating this without making it sound like a vector? Collection, sequence, e.g.?

File: spec.md
Comment (seldridge) [inline spec.md:657]: This is a bit of a hard transition from the previous paragraph. I thought this was continuing the example above.

Perhaps this would be better located at the start of the sub-section.

File: spec.md
Comment (seldridge) [inline spec.md:670]: Hemmingway speaking from the grave:

```suggestion
Each variant is named with a tag.
Each variants has a passive type (see [@sec:passive-types]).
```

File: spec.md
Comment (seldridge) [inline spec.md:724]:
```suggestion
`Probe`{.firrtl} and `RWProbe`{.firrtl} are created using the `probe`{.firrtl} and `rwprobe`{.firrtl} expressions, respectively (see [@sec:probes]).
```

File: spec.md
Comment (seldridge) [inline spec.md:977]: I would avoid discussions of a FIRRTL compiler and opt instead for language around legal and illegal FIRRTL circuits. Also, while there is de facto one ("the") FIRRTL compiler, there may be many.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:555]: > Ground types are types which are not composed of other types.
> They are the integer types, clocks, and resets.

Reworded to this. 

I made a note of the ground type terminology. We can revisit this.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:573]: Agreed. There was a whole paragraph on this, and I stripped it out. 

I'll put it in my notes to add to the ABI doc and remove this TODO.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:627]: > A bundle type consists of multiple fields.

Changed.

File: spec.md
Comment (seldridge) [inline spec.md:571]: Using `pad` with a reference to the `pad` primop may be more accurate than saying `extend` which could mean either zero-extend or sign-extend (and FIRRTL doesn't have both operations, just one operation that is sensitive to the type).

As commentary (either in the PR or future), interpreting sign-extension of a zero-width signed integer as adding leading zeros was something controversial when I tried to land this on upstream MLIR. There were discussions about whether this should _actually add ones_. I.e., the value of a zero-width signed integer is actually `-1`. The end result was that this was undefined and MLIR will error if you try to sign extend a zero-width signed integer.

This is again in the Bruce Hornsby bucket and it's fine that FIRRTL does it this way. A reader with experience (or a mathematical background) may do a double-take on these sentences.

Did you have a similar reaction given your math background?

File: spec.md
Comment (mwachs5) [inline spec.md:550]: you seem to have removed analog from the ground types list below, but not added them to this list... where do they go in this classification?

Comment (mmaloney-sf) [reply spec.md:550]: I need to ask some questions about analog types to decide where they should go.

Related to that is what "ground" type even means. The way we classify types is really lumpy, and I want to feel things out a bit.

**Comment by mmaloney-sf** [comment]
@darthscsi I rewrote the Constant Types section as best as I could.

Being such a massively cross-cutting feature, it's hard to spec out in just one section. I think the rewritten version touches on all points from the original:

- include a (tentative) rationale for the feature
- remove the example of a bundle with a nested `const` -- it's theoretically interesting, but not really speced out, so let's not draw attention to it
- mention that `const` doesn't mean `static` -- these things might be unknown because of `input` ports
- explain the rules everyone will agree one (literals are `const`, prim ops of `const` are `const`, muxes of `const` are `const`).
- Gesture at how `const T` can be implicitly coerced/cast to `T`.
- Gesture at the connect rules for `const` (source must be `const`, any conditionals must have constant conditions)
-  Reword "indexing" into aggregates in terms of subcomponents (I haven't written the subcomponents section yet, but bear with me please!)
- Moved "A note on implementation" to the ABI doc.

Thank you @seldridge for letting me bounce my thoughts off of you.

**Comment by mmaloney-sf** [comment]
One big addition at the last minute:

I define a **Connectable Type** as a type which may appear as the type of an expression which may participate in a `connect` statement.

Connectable types appear internally in the FIRRTL MLIR dialect. They affect how components are split up when lowering to Verilog. 

Currently, I only rewrote the enumerations section in terms of these. (Namely, the types of variants must be connectable and passive). But as I get to the section on Connections, we can apply this definition cleanly there.

This definition will also help us untangle some thorny issues in FIRRTL, which add a disproportionate amount of complexity to the spec:

- what it means for non-hardware to be packaged in aggregates (eg, bundles of probes or properties), and
- how we should think of the `Analog` type

**Comment by mmaloney-sf** [comment]
I forgot -- I'm a bit stuck on the section on Probes and Type Inference.

@dtzSiFive Can you advise me on how to handle this section?

File: spec.md
Comment (mwachs5) [reply spec.md:575]: sorry how was this resolved?

File: spec.md
Comment (mwachs5) [inline spec.md:734]: this seems like a possible mischaracterization... in firrtl there is no concept of hierarchical path to a component in a design (except in targets...), in verilog there is. We could have just implemented probes like that in the firrtl spec, the advantage isn't really that you do or do not know the hierarchical path to the component in the design.  I'm not sure that is what makes them useful for testing and verification, they are useful in verilog because there is a sense that the verilog hierarchy is "real" and XMRs are "not real" -- you can add XMRs without adding any "real" ports. Would love @seldridge, @darthscsi  to weigh in on this rewrite.

File: spec.md
Comment (mwachs5) [inline spec.md:737]: is this an intentional change to the spec? Because the example you have deleted is not a ground type. If you are trying to make a functional change to the spec that probes have to be ground types, please make a different PR for that.

**Review by mwachs5 — DISMISSED** [review]
it seems this PR is (unintentionally?) making probes more restricted to be only on ground types. If that is the intention please break that change to its own, non-patch PR.

File: spec.md
Comment (seldridge) [inline spec.md:580]: Also, analog?

File: spec.md
Comment (seldridge) [inline spec.md:623]: This is fine as commentary here. However, there are a few things missing/unclear: (1) How is this represented in the ABI? (There is a missing ABI section on Analog lowering.) (2) Given the ABI, Analog does lower to `inout`, therefore, is it appropriate to say `Analog` is "similar to how inout ports are used in Verilog"?

File: spec.md
Comment (seldridge) [inline spec.md:668]: I'd opt for a "collection of types" as opposed to "collection of values".

File: spec.md
Comment (seldridge) [inline spec.md:659]: Mega-nit: FWIW, the formatter is going to add a space after triple-backtick. It's fine to just handle this with a format PR.

File: spec.md
Comment (seldridge) [inline spec.md:702]: Nit: the ellipsis is unnecessary. You could make it clearly irrelevant by making `Processor` an `extmodule`.

File: spec.md
Comment (seldridge) [reply spec.md:734]: A major use case of probes is that you can use them without knowing the hierarchical path which makes them work for separate compilation where annotations are intractable. They are highly amenable to compiler implementations because they work with hierarchy manipulations and optimizations (e.g., constant propagation) whereas annotations were a hug pain here.

That said, drawing a comparison to something which doesn't exist in FIRRTL (a hierarchical path) may not be the right way to explain this. I.e., they are useful for verification because they let you read from circuit components without creating ports. (This is restating Megan's penultimate sentence.)

File: spec.md
Comment (seldridge) [reply spec.md:737]: I thought this is how these are being implemented from Chisel now. Probes of aggregates don't work well with the the Port Lowering ABI v1 (i.e., `LowerTypes`) on public modules. Hence, this is a restriction that I think is already in place.

We're at a `[major]` spec boundary given that the spec already has public modules. So, I'm fine if this is just a `[major]` change.

File: spec.md
Comment (seldridge) [inline spec.md:766]: Nit:
```suggestion
Probes are generally lowered to hierarchical names in Verilog.
```

File: spec.md
Comment (seldridge) [inline spec.md:913]:
```suggestion
The bit widths are given in angle brackets, such as the `8`{.firrtl} in `UInt<8>`{.firrtl}.
```

File: spec.md
Comment (seldridge) [inline spec.md:917]: The copy here introduces `uninferred` for the first time. However, `inferred` was introduced in the discussion of types. Does it make sense to introduce `uninferred` earlier as well?

File: spec.md
Comment (seldridge) [inline spec.md:934]: Supreme nit:
```suggestion
The uninferred `Reset`{.firrtl} type will be inferred to either a synchronous reset, `UInt<1>`{.firrtl}, or to an asynchronous reset, `AsyncReset`{.firrtl}.
```

File: spec.md
Comment (seldridge) [inline spec.md:944]: This sentence and the previous one seem to restate the same point. I think this is actually incorrectly repeated twice.

File: spec.md
Comment (seldridge) [inline spec.md:960]: This example seems like a non-sequitor?

File: spec.md
Comment (seldridge) [inline spec.md:966]: Mega-nit: 
```suggestion
1. Otherwise, the reset is inferred as synchronous. (I.e., the uninferred reset is only invalidated or is driven by or drives only synchronous resets.)
```

File: spec.md
Comment (seldridge) [inline spec.md:969]: Does the PR want to call out what are the casts in the example below?

File: spec.md
Comment (seldridge) [inline spec.md:1015]: I'm not sure if type coercion is the right concept here. Type equivalence is really this concept that is used narrowly to check the legality of statements or expressions. Put differently, the type equivalence relation says that `UInt<8>` is equal to `UInt`. This really only matters for connects to allow:

```
wire a: UInt
wire b: UInt<8>
connect b, a
```

However, the _type_ of that connect is still `(UInt<8>, UInt) -> ()`. I.e., the type of `a` is not coerced in any way. `a` is still an uninferred `UInt`.

Put differently again, this entire section is basically, "How do I write a verifier for connect ops?" as opposed to anything type theoretic.

It might be possible to reframe all this as "connects require exact type equivalence and uninferred is coereced to its inferred version". However, that is practically not what is happening.

File: spec.md
Comment (seldridge) [inline spec.md:1023]: This sentence doesn't seem appropriate for type equivalence.

File: spec.md
Comment (seldridge) [reply spec.md:1015]: Maybe this entire section of type equivalence should really be moved under connects and called "connect legality" or "connect verification". 🤔

File: spec.md
Comment (mmaloney-sf) [reply spec.md:575]: `Analog` was simply moved back to the Ground Types subsection.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:737]: This is the kind of thing I'm looking to flush out with this spec work.

I'm fine with making this a `[major]` change. But let's be sure not to release a new version until the dust has all settled.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:659]: Wait what? Have I been writing markdown codeblocks wrong my whole life?

I'll submit a separate PR to run the formatter.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:766]: Fixed.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:913]: `{👼}`

File: spec.md
Comment (mwachs5) [reply spec.md:737]: sure i am not objecting to a spec change, just seems like that should be done broken out from a big rewrite of things, given that its like a one sentence change and deletion of an example to say that probes don't work on aggregates.

I am not making a statement on how chisel is implementing them now/ near future, defer to @dtzSiFive on that.

Comment (mmaloney-sf) [reply spec.md:737]: @mwachs5 I promoted the PR to `[major]`.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:934]: I intentionally dropped the commas, since otherwise it could be read a bit ambiguously.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:1023]: This was the existing copy. It's in rough shape.

I want to make a separate PR at some point to rename "type equivalence" to something more sane.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:969]: This was the original copy.

The entire notion of type inference (width and reset) are borked in the spec.

I'm not even sure it's worth the time to try and rewrite it until we have a more principled approach to handling it.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:966]: This was the original copy.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:1015]: I had originally considered rewriting this section, but it will need some dedicated attention.

I'll change this intro back to the original copy.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:960]: This was the original copy.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:944]: This was the original copy.

File: spec.md
Comment (mmaloney-sf) [reply spec.md:596]: This section is the original copy.

I'll weigh in on the `SInt<0>` battle at some point. But not today :)

File: spec.md
Comment (mmaloney-sf) [reply spec.md:623]: This is the original copy.

File: spec.md
Comment (mmaloney-sf) [reply spec.md]: Rewrote this section after consulting @dtzSiFive.

It's still hand-wavy, but only due to the fact that all of type inference is one big hand-wave at this point. It should suffice for now.

File: spec.md
Comment (mmaloney-sf) [reply spec.md]: I'm flipping this back to `[patch]` and removing this line.

This point represents a pretty big gap. But let me leave this without the restriction, and we can address it after all the low-hanging fruit is gone.

Comment (mwachs5) [reply spec.md]: thanks :) nit I see you reverted the sentence but you didn't revert the example, which is now just sort of ambiguous and confusing 😂 If you think this is a noncontroversial change can just open a PR right now to remove it, no need to block on this larger (in lines of text) change

File: spec.md
Comment (mmaloney-sf) [reply spec.md]: Removed the references to hierarchical paths.

I did slip one usage of the term "physical hardware". Please consider this an informal statement. We will work towards a formalization of what this means before too long.

**Pull request merged**
